Fix: safety checks type (#785)
trezor/connect: docs/protobuf-type-definitions.md

`trezor-common` is included in Connect as a git submodule mounted at `submodules/trezor-common`.

Here, `.proto` definitions are translated to a JSON format using [proto2js](https://www.npmjs.com/package/proto2js) package. This JSON is used on runtime by the [trezor-link](https://github.com/trezor/trezor-link/) (de)serialization logic and to generate the Flow and Typescript definitions.

The JSON is transformed to Flow and/or TypeScript definitions by a script in `scripts/protobuf-types.js`. The script also applies 'patches' I.e. after-the-fact fixes manually described in `scripts/protobuf-patches.js`. The patches compensate for/fix
- The source `.proto` definitions that do not reflect the actual business logic. Usually fields marked as required which are in fact optional.
- Fields typed as `uint32`, `uint64`, `sint32`, `sint64` in protobuf that need to be represented as strings in runtime because of javascript number's insufficient range. Runtime conversion is handled automatically by `trezor-link`.

> - Similarly, fields typed as `bytes` in protobuf that are represented as base64 string in runtime.

Reviewer: ```suggestion
- Similarly, fields typed as `bytes` in protobuf may be represented as hexadecimal `string`, `Buffer`, `Uint8Array` or `Array<number>` in runtime.
```

see
https://github.com/trezor/connect/blob/develop/scripts/protobuf-patches.js#L111
or
https://github.com/trezor/connect/blob/develop/scripts/protobuf-patches.js#L133-L134
or
https://github.com/trezor/connect/blob/develop/scripts/protobuf-types.js#L20

Author: Ah, so this is even wider. Thanks for linking the examples!

Suggestion applied [here](https://github.com/trezor/connect/pull/785/commits/e8efb21fe44d752b3e76a151e7eb661a46502087).

# From Protobuf to TypeScript and Flow

This document describes how the definitions of protobuf messages maintained in the
firmware repo are semi-automatically translated to Flow and TypeScript definitions that are used in Connect codebase and by Connect users respectively.

## The Pipeline

> Firmware team maintains the `.proto` definitions [in the firmware repo](https://github.com/trezor/trezor-firmware/tree/master/common/protob) and duplicates them in the [trezor-common](https://github.com/trezor/trezor-common) repository.

Reviewer: `.proto` definitions are maintained in the firmware repository and exported in to the trezor-common repository as a read only copy.

Author: Edited. Pls check.

Reviewer: once you are using word "repo" and another time "repository".
what about 3rd party contributors that maintain protobuf messages and they are not members of firmware team? (like for example cardano)
that is why i've suggested this change

Author: > once you are using word "repo" and another time "repository".

Fixed. Thx.

> what about 3rd party contributors that maintain protobuf messages and they are not members of firmware team? 

Good point. It didn't appear to me at first sight (still didn't get used to working in the open) and I wanted the text to be more personal and less passive/technical.

Hopefully fixed now. .)

`trezor-common` is included in Connect as a git submodule mounted at `submodules/trezor-common`.

> These `.proto` definitions are translated to a JSON format using [proto2js](https://www.npmjs.com/package/proto2js) package. This JSON is used on runtime by the [Link](https://github.com/trezor/trezor-link/) (de)serialization logic and to generate the Flow and Typescript definitions.

Reviewer: ```suggestion
`.proto` definitions are translated to a JSON format using [proto2js](https://www.npmjs.com/package/proto2js) package. This JSON is used on runtime by the [trezor-link](https://github.com/trezor/trezor-link/) (de)serialization logic and to generate the Flow and Typescript definitions.
```

Author: Edited as suggested. Resolving.

The JSON is transformed to Flow and/or TypeScript definitions by a script in `scripts/protobuf-types.js`. The script also applies 'patches' I.e. after-the-fact fixes manually described in `scripts/protobuf-patches.js`. The patches compensate for/fix
- The source `.proto` definitions that do not reflect the actual business logic. Usually fields marked as required which are in fact optional.

> - Fields that are transferred as numbers in protobuf but need to be represented as strings in JS because JS's number doesn't have sufficient range. Runtime conversion is handled automatically by Link.

Reviewer: ```suggestion
- Fields expected as `uint32/uint64` and `sint32/sint64` in protobuf while in fact they are represented as strings because of javascript number insufficient range. Runtime conversion is handled automatically by `trezor-link`.
- Fields expected as `bytes` in protobuf  while in fact they are represented as hex strings. Runtime conversion is handled automatically by `trezor-link`.
```

Author: Edited. Pls check. .)

Reviewer: you didn't mentioned suggested `bytes` as hex string

Author: > you didn't mentioned ...

You're right. Added that as well.

> ... suggested ...

though I can't see where it was suggested? Maybe I'm missing something? o.O

> - Optional protobuf fields that get typed as `<T> | undefined` but are in fact deserialized as `<T> | null`. This could be handled globally by Link. The patches exist mainly for historical reasons.

Reviewer: ```suggestion
- Optional protobuf fields that get typed as `<T> | undefined` but are in fact deserialized as `<T> | null`. This could be handled globally by `trezor-link`. The patches exist mainly for historical reasons.
```

Author: Edited as suggested. Resolving.

All these steps are done manually and all the generated files are tracked in git. It's also not uncommon to circumvent
some step by eg. generating the messages.json file not from the Common submodule but directly from the firmware repo.

## Example

```bash
# Clone the connect repo.
git clone git@github.com:trezor/connect.git
cd connect
# Fetch all the submodules.
# Technically, `--recursive` isn't needed to fetch .proto files.
# But better be defensive.

> git submodule update --init --recursive

Reviewer: ```suggestion
make submodules
```

Author: Edited as suggested. Resolving.
